KyleHouston/memory-game.js

```diff
@@ -0,0 +1,119 @@
+var MemoryGame = (function() {
+
+	function GameCtor(cardset) {
+		var slots, //values of shuffled cards;
+				//sparse array: will have elements deleted as cards are removed
+			length,//total slots, including gaps
+			there; //position of face-up card if any, or false
+
+		// Helper functions which need access to closure vars;
+		//  some fns will be made public as instance methods:
+		var reset = function() {
+			slots = cardset.values();
+			length = slots.length;
+			there = false;
+			shuffle(slots);
+		}
+		reset();// reset now as part of init'ing
+
+		var remainsAt = function(where) {//--> boolean
+			return slots[where]!==undefined;
+		}
+		var valueAt = function(where) {//--> card val
+			return slots[where];
+		}
+		var removeAt = function(where) {
+			delete slots[where];
+		}
+		var faceupValue = function() {//--> card val
+			return valueAt(there);
+		}
+		var faceupWhere = function() {//--> integer
+			return there;
+		}
+		var remaining = function() {//--> array of integers
+			return Object.keys(slots).map(Number);
+		}
+
+		var lift = function(here) {//--> display string
+			if (!isValid(here,length)) return false;
+			if (!remainsAt(here)) return false;
+			if (there===here) return false;
+
+			// must be a face-down card here; proceed...
+			var valHere = valueAt(here);
+			var fileName = cardset.display(valHere); //display card face file name for debugging
```

Comment:
Because this is ultimately returning the capitalized version of the card name (eg `Ace_of_Hearts.svg` instead of `ace_of_hearts.svg`), images for card faces aren't displaying, at least on my machine. You could approach this in at least two different ways: by having your code automatically downcase the image names, or by manually capitalizing the names of the image files to match.